go/extractor/cli/go-autobuilder/go-autobuilder.go

```diff
@@ -355,240 +354,210 @@ func main() {
-				log.Printf("Moving %s/%s to %s/%s.\n", srcdir, file, scratch, file)
-				err := os.Rename(filepath.Join(srcdir, file), filepath.Join(scratch, file))
-				if err != nil {
-					log.Fatalf("Failed to move file %s to the temporary directory: %s\n", file, err.Error())
-				}
+	files, err := dir.Readdirnames(-1)
+	if err != nil {
+		log.Fatalf("Failed to read source directory %s: %s\n", srcdir, err.Error())
+	}
+	for _, file := range files {
+		if file != filepath.Base(scratch) {
+			log.Printf("Moving %s/%s to %s/%s.\n", srcdir, file, scratch, file)
+			err := os.Rename(filepath.Join(srcdir, file), filepath.Join(scratch, file))
+			if err != nil {
+				log.Fatalf("Failed to move file %s to the temporary directory: %s\n", file, err.Error())
 			}
 		}
+	}
 
-		// create a new folder which we will add to GOPATH below
-		// Note we evaluate all symlinks here for consistency: otherwise os.Chdir below
-		// will follow links but other references to the path may not, which can lead to
-		// disagreements between GOPATH and the working directory.
-		realSrc, err := filepath.EvalSymlinks(srcdir)
-		if err != nil {
-			log.Fatalf("Failed to evaluate symlinks in %s: %s\n", srcdir, err.Error())
-		}
+	// create a new folder which we will add to GOPATH below
+	// Note we evaluate all symlinks here for consistency: otherwise os.Chdir below
+	// will follow links but other references to the path may not, which can lead to
+	// disagreements between GOPATH and the working directory.
+	realSrc, err := filepath.EvalSymlinks(srcdir)
+	if err != nil {
+		log.Fatalf("Failed to evaluate symlinks in %s: %s\n", srcdir, err.Error())
+	}
 
-		root := filepath.Join(realSrc, "root")
+	root := filepath.Join(realSrc, "root")
 
-		// move source files to where Go expects them to be
-		newdir := filepath.Join(root, "src", importpath)
-		err = os.MkdirAll(filepath.Dir(newdir), 0755)
-		if err != nil {
-			log.Fatalf("Failed to create directory %s: %s\n", newdir, err.Error())
-		}
-		log.Printf("Moving %s to %s.\n", scratch, newdir)
-		err = os.Rename(scratch, newdir)
-		if err != nil {
-			log.Fatalf("Failed to rename %s to %s: %s\n", scratch, newdir, err.Error())
-		}
+	// move source files to where Go expects them to be
+	newdir := filepath.Join(root, "src", importpath)
+	err = os.MkdirAll(filepath.Dir(newdir), 0755)
+	if err != nil {
+		log.Fatalf("Failed to create directory %s: %s\n", newdir, err.Error())
+	}
+	log.Printf("Moving %s to %s.\n", scratch, newdir)
+	err = os.Rename(scratch, newdir)
+	if err != nil {
+		log.Fatalf("Failed to rename %s to %s: %s\n", scratch, newdir, err.Error())
+	}
 
-		// schedule restoring the contents of newdir to their original location after this function completes:
-		defer restoreRepoLayout(newdir, files, filepath.Base(scratch), srcdir)
+	// schedule restoring the contents of newdir to their original location after this function completes:
+	defer restoreRepoLayout(newdir, files, filepath.Base(scratch), srcdir)
```

Comment:
Moving defer into a different function means it will run at a different time. Instead you'll need to do something like return a cleanup lambda that the caller defers until the end of the autobuilder process.

Reply:
That approach isn't ideal because we do lots of calls to `log.Fatalf` in between. I've made the refactoring a bit more fine-grained instead. I think the result is still quite concise and readable.